src/dp_nat.c

```diff
@@ -296,6 +297,184 @@ void dp_nat_chg_ip(struct dp_flow *df, struct rte_ipv4_hdr *ipv4_hdr,
 	}
 }
 
+static void dp_calculate_icmp_checksum(struct rte_icmp_hdr *icmp_hdr, size_t icmp_len)
+{
+	uint32_t sum = 0;
+	uint16_t word;
+	uint8_t *ptr = (uint8_t *)icmp_hdr;
+
+	icmp_hdr->icmp_cksum = 0;
+
+	for (size_t i = 0; i < icmp_len; i += 2) {
+		word = (uint16_t)((uint16_t)ptr[i] | ((uint16_t)ptr[i + 1] << 8));
+		sum += word;
+	}
+
+	while (sum >> 16)
+		sum = (sum & 0xFFFF) + (sum >> 16);
+
+	icmp_hdr->icmp_cksum = ~((uint16_t)sum);
+}
+
+int dp_nat_chg_ipv6_to_ipv4_hdr(struct dp_flow *df, struct rte_mbuf *m, uint32_t nat_ip, rte_be32_t *dest_ip /* out */)
+{
+	struct rte_ether_hdr *eth_hdr;
+	struct rte_ipv6_hdr *ipv6_hdr;
+	struct rte_ipv4_hdr *ipv4_hdr;
+	struct rte_icmp_hdr *icmp_hdr;
+	struct rte_udp_hdr *udp_hdr;
+	struct rte_tcp_hdr *tcp_hdr;
+	uint8_t l4_proto;
+
+	eth_hdr = rte_pktmbuf_mtod(m, struct rte_ether_hdr *);
+	eth_hdr->ether_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+
+	ipv6_hdr = dp_get_ipv6_hdr(m);
+	*dest_ip = *(int *)&ipv6_hdr->dst_addr[12];
+	l4_proto = ipv6_hdr->proto;
+
+	// Adjust the packet data to fit IPv4
+	if (rte_pktmbuf_adj(m, sizeof(struct rte_ipv6_hdr) - sizeof(struct rte_ipv4_hdr)) == NULL)
+		return DP_ERROR;
+
+	// Move the Ethernet header to just before the IPv4 header
+	memmove(rte_pktmbuf_mtod(m, uint8_t *), eth_hdr, sizeof(struct rte_ether_hdr));
+
+	// Setup the new IPv4 header
+	ipv4_hdr = dp_get_ipv4_hdr(m);
+	ipv4_hdr->version_ihl = 0x45;
+	ipv4_hdr->type_of_service = (uint8_t)((ntohl(ipv6_hdr->vtc_flow) >> 20) & 0xFF);
+	ipv4_hdr->total_length = rte_cpu_to_be_16(rte_pktmbuf_data_len(m) - (uint16_t)sizeof(struct rte_ether_hdr));
+	ipv4_hdr->packet_id = 0;
+	ipv4_hdr->fragment_offset = 0;
+	ipv4_hdr->time_to_live = ipv6_hdr->hop_limits;
+	ipv4_hdr->next_proto_id = l4_proto;
+	ipv4_hdr->src_addr = htonl(nat_ip);
+	ipv4_hdr->dst_addr = *dest_ip;
+	ipv4_hdr->hdr_checksum = 0;
+
+	m->tx_offload = 0;
+	m->packet_type &= ~RTE_PTYPE_L3_MASK;
```

Comment:
At first glance this seems strange to me, but I would have to check it via `dpservice-dump` which is the only thing that relies on this value to be right. If you have a test at hand can you look if the output still makes sense after this? `dp_virtsvc.c` does this in a more complicated way.

Reply:
Ok. When you now say it, I checked the dp_virtsvc.c and the realized that I am missing the line
`m->packet_type |= RTE_PTYPE_L3_IPV4;`
It is even better to combine everything to a single line so I would go with the dp_virtsvc approach.
`m->packet_type = (m->packet_type & ~RTE_PTYPE_L3_MASK) | RTE_PTYPE_L3_IPV4;`